File: 总题库/974.SubarraySumsDivisiblebyK.py
class Solution:
    def subarraysDivByK(self, A: List[int], K: int) -> int:
        record =  {0:1}
        curSum = 0
        ans = 0

        for number in A:
            curSum += number
            modulus = (curSum % K + K) % K
            if modulus not in record:
                record[modulus] = 1
            else:
                record[modulus] += 1

        for key in record:
            ans += record[key] * (record[key] - 1) // 2

        return ans

File: 总题库/974.SubarraySumsDivisiblebyK.cpp
class Solution {
public:
    int subarraysDivByK(vector<int>& A, int K) {
        unordered_map<int, int> record = {{0, 1}};
        int sum = 0, ans = 0;
        for (int i : A) {
            sum += i;
            int modulus = (sum % K + K) % K;
            ++record[modulus];
        }

        for (auto [x, cx]: record) {
            ans += cx * (cx - 1) / 2; // 选出两个的组合；
        }
        return ans;
    }
};